Scripts/script-GetDuplicatesMl.yml

```diff
@@ -0,0 +1,641 @@
+
+          return False
+
+      @staticmethod
+      def get_minimal_edit_distance(group1, group2):
+          min_value = None
+          for x in group1:
+              for y in group2:
+                  ed = editdistance.eval(x, y)
+                  if min_value is None or ed < min_value:
+                      min_value = ed
+
+          return min_value
+
+      @staticmethod
+      def intersection_set(x, y):
+          return set.intersection(*[set(x), set(y)])
+
+      @staticmethod
+      def union_set(x, y):
+          return set.union(*[set(x), set(y)])
+
+      @staticmethod
+      def get_hashable_from_dict(d):
+          return [(k, v) for k, v in d.items() if isinstance(v, collections.Hashable)]
+
+
+      @staticmethod
+      def jaccard_similarity(x, y):
+          if x is None or y is None:
+              return 0
+
+          if type(x) == dict:
+              x = Utils.get_hashable_from_dict(x)
+          if type(y) == dict:
+              y = Utils.get_hashable_from_dict(y)
+
+          x = [v for v in x if isinstance(v, collections.Hashable)]
+          y = [v for v in y if isinstance(v, collections.Hashable)]
+
+          if len(x) == 0 or len(y) == 0:
+              return 0
+
+          intersection_cardinality = len(Utils.intersection_set(x, y))
+          union_cardinality = len(Utils.union_set(x, y))
+          return intersection_cardinality / float(union_cardinality)
+
+  class IncidentFeatures:
+      def __init__(self, incident1, incident2):
+
+          self.incident1 = incident1
+          self.incident2 = incident2
+
+          self.indicators1 = incident1['indicators']
+          self.indicators2 = incident2['indicators']
+
+          self.labels_map1 = Utils.get_incident_labels_map(self.incident1['labels'])
+          self.labels_map2 = Utils.get_incident_labels_map(self.incident2['labels'])
+
+          domains1 = Utils.get_unique_list(
+              self.indicators1.get('Domain', []) + Utils.get_domains(self.indicators1, self.labels_map2))
+          domains2 = Utils.get_unique_list(
+              self.indicators2.get('Domain', []) + Utils.get_domains(self.indicators2, self.labels_map2))
+
+          if len(domains1) > 0:
+              self.indicators1['Domain'] = domains1
+          if len(domains2) > 0:
+              self.indicators2['Domain'] = domains2
+
+      def get_email_features(self):
+          emails1, emails2 = self.indicators1.get('Email', []), self.indicators2.get('Email', [])
+          if Utils.is_groups_cross_empty(emails1, emails2):
+              return {}
+
+          email_user_names1 = map(Utils.extract_email_username, emails1)
+          email_user_names2 = map(Utils.extract_email_username, emails2)
+
+          features = Utils.calculate_similar_pattern('email_username', email_user_names1, email_user_names2)
+
+          features['email_min_ld'] = Utils.get_minimal_edit_distance(email_user_names1, email_user_names2)
+
+          return features
```

Comment:
Try to look at numpy maybe for those array manipulation

Reply:
V

Reply:
V